File: javascript/json-transform/src/functions/TransformerFunctionRange.ts
import TransformerFunction from "./common/TransformerFunction";
import { ArgType } from "./common/ArgType";
import FunctionContext from "./common/FunctionContext";
import JsonElementStreamer from "../JsonElementStreamer";
import { createAsyncSequence } from "@wortise/sequency";
import { RoundingModes } from "./common/FunctionHelpers";
import BigNumber from "bignumber.js";

class TransformerFunctionRange extends TransformerFunction {
  constructor() {
    super({
      arguments: {
        start: { type: ArgType.BigDecimal, position: 0, defaultIsNull: true },
        end: { type: ArgType.BigDecimal, position: 1, defaultIsNull: true },
        step: { type: ArgType.BigDecimal, position: 2, defaultBigDecimal: 1 },
      },
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    let start: BigNumber | null;
    let end: BigNumber | null;
    let step: BigNumber;
    if (context.has("start")) {
      start = await context.getBigDecimal("start");
      end = await context.getBigDecimal("end");
      step = (await context.getBigDecimal("step")) ?? BigNumber(1);
    } else {
      const arr = await context.getJsonArray(null);
      if (!arr) {
        return null;
      }
      start = arr[0];
      end = arr[1];
      step = arr[2] ?? BigNumber(1);
    }
    // sanity check
    if (
      start === null ||
      end === null ||
      (end.lt(start) && step.isPositive()) ||
      (end.gt(start) && step.isNegative())
    ) {
      return null;
    }
    const size = end.minus(start).dividedToIntegerBy(step).plus(1).integerValue(RoundingModes.DOWN).toNumber();

    let value = start;
    return JsonElementStreamer.fromTransformedStream(
      context,
      createAsyncSequence({
        next: () => {
          const result = value;
          value = value.plus(step);
          return Promise.resolve({ value: result });
        },
      }).take(size),
    );
  }
}

export default TransformerFunctionRange;

File: java/json-transform/src/main/java/co/nlighten/jsontransform/functions/TransformerFunctionRange.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.JsonElementStreamer;
import co.nlighten.jsontransform.functions.common.*;

import java.math.BigDecimal;
import java.util.Map;
import java.util.concurrent.atomic.AtomicReference;
import java.util.stream.Stream;

public class TransformerFunctionRange extends TransformerFunction {
    public TransformerFunctionRange() {
        super(FunctionDescription.of(
            Map.of(
            "start", ArgumentType.of(ArgType.BigDecimal).position(0),
            "end", ArgumentType.of(ArgType.BigDecimal).position(1),
            "step", ArgumentType.of(ArgType.BigDecimal).position(2).defaultBigDecimal(BigDecimal.ONE)
            )
        ));
    }
    @Override
    public Object apply(FunctionContext context) {
        BigDecimal start;
        BigDecimal end;
        BigDecimal step;
        if (context.has("start")) {
            start = context.getBigDecimal("start");
            end = context.getBigDecimal("end");
            step = context.getBigDecimal("step");
        } else {
            var arr = context.getJsonArray(null);
            var adapter = context.getAdapter();
            var length = adapter.size(arr);
            start = length > 0 ? adapter.getNumberAsBigDecimal(adapter.get(arr, 0)) : null;
            end = length > 1 ? adapter.getNumberAsBigDecimal(adapter.get(arr, 1)) : null;
            step = length > 2 ? adapter.getNumberAsBigDecimal(adapter.get(arr, 2)) : BigDecimal.ONE;
        }
        // sanity check
        if (start == null ||
            end == null ||
            (end.compareTo(start) < 0 && step.signum() > 0) ||
            (end.compareTo(start) > 0 && step.signum() < 0)
        ) {
            return null;
        }
        var size = end.subtract(start).divideToIntegralValue(step).add(BigDecimal.ONE).intValue();
        var value = new AtomicReference<>(start);
        return JsonElementStreamer.fromTransformedStream(context,
                Stream.generate(() -> value.getAndUpdate(next -> next.add(step))).limit(size)
        );
    }
}
